### PR TITLE
XLS(X) Reading

### PR DESCRIPTION
**Pull request by kustosz**
### Pull Request Description
It's purpose in life is to read XLS(X) files.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/main/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/main/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/main/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/main/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

File: distribution/lib/Standard/Table/0.1.0/src/Io/Spreadsheet.enso [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+     Read the active sheet of an XLS from disk and convert it into a table.
+
+         import Standard.Table
+         import Standard.Examples
+
+         example_xlsx_to_table = Table.from_xls Examples.xls
+
+   > Example
+     Read the sheet named `Dates` from an XLS and convert it to a table.
+         import Standard.Table
```
Comment (iamrecursion) [inline distribution/lib/Standard/Table/0.1.0/src/Io/Spreadsheet.enso]:
```suggestion
     Read the sheet named `Dates` from an XLS and convert it to a table.
     
         import Standard.Table
```

File: distribution/lib/Standard/Table/0.1.0/src/Io/Spreadsheet.enso [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+         import Standard.Examples
+
+         example_xls_to_table = Table.from_xls Examples.xls sheet='Dates'
+from_xls : File.File -> Integer | Text | Nothing -> Boolean -> Text -> Table
```
Comment (iamrecursion) [inline distribution/lib/Standard/Table/0.1.0/src/Io/Spreadsheet.enso:144]:
```suggestion
         example_xls_to_table = Table.from_xls Examples.xls sheet='Dates'
         
from_xls : File.File -> Integer | Text | Nothing -> Boolean -> Text -> Table
```

**Review by iamrecursion — APPROVED** [review]
If we want to commit binary data to the repository like this we should probably think about enabling LFS or similar. Doing binary data as normal files isn't great for git.

File: distribution/lib/Standard/Table/0.1.0/src/Io/Spreadsheet.enso [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+   > Example
+     Read the sheet named `Dates` from an XLSX and convert it to a table.
+         import Standard.Table
+         import Standard.Examples
+
+         example_xlsx_to_table = Examples.xlsx.read_xlsx sheet='Dates'
+File.File.read_xlsx : Integer | Text | Nothing -> Boolean -> Text -> Table
+File.File.read_xlsx sheet=Nothing has_header=True prefix='C' =
+    here.from_xlsx this sheet has_header prefix
```
[the remark this answers is not in the document]
Comment (radeusgd) [reply distribution/lib/Standard/Table/0.1.0/src/Io/Spreadsheet.enso:40]: I second this

File: std-bits/table/src/main/java/org/enso/table/format/xlsx/Reader.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+   * @param hasHeaders specifies whether the first non-empty row of the sheet should be used for
+   *     column names.
+   * @param unnamedColumnPrefix specifies the prefix to use for missing columns.
+   * @param mkDate a function converting java-based dates into a format understandable by the
+   *     caller.
+   * @return a {@link Table} containing the specified data.
+   * @throws IOException when the input stream cannot be read.
+   */
+  public static Table read_xlsx(
+      InputStream is,
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/format/xlsx/Reader.java]: I think our code guide advises against such short names, unless they are stuff like iterators (`i`, `j` etc.), so maybe `inputStream` will work here?

File: std-bits/table/src/main/java/org/enso/table/format/xlsx/Reader.java [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+      var headerRow = sheet.getRow(minRow);
+      minRow++;
+      for (int i = minCol; i < maxCol; i++) {
+        var cell = headerRow.getCell(i, Row.MissingCellPolicy.RETURN_BLANK_AS_NULL);
+        if (cell == null || cell.getCellType() != CellType.STRING) {
+          colNames.add(unnamedColumnPrefix + (i - minCol));
+        } else {
+          colNames.add(cell.getStringCellValue());
+        }
+      }
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/format/xlsx/Reader.java:122]: I think our dataframes do not deal well with non-unique column names, so unless we _do want_ to support this (in which case it would be good to create an issue tracking that, because currently many operations can lead to undefined behaviour in presence of name clashes in table columns), I think we should do something here to deduplicate these column names.

Comment (radeusgd) [reply std-bits/table/src/main/java/org/enso/table/format/xlsx/Reader.java:122]: I'm not sure what is the case with CSV loading, but if it is the same thing maybe it should also get some deduplication? Or at least some kind of a warning (although that will not show up in the IDE, at least for now). Because I think this can lead to some very confusing situations and I'm sure there are CSVs/Excel sheets with duplicates exist in the wild.

File: test/Table_Tests/src/Spreadsheet_Spec.enso [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+            t = read_method file sheet=1
+            t.columns.map .name . should_equal ['Name', 'Quantity', 'Price']
+            t.at 'Name' . to_vector . should_equal ['blouse', 't-shirt', 'trousers', 'shoes', 'skirt', 'dress']
+            t.at 'Quantity' . to_vector . should_equal [10, 20, Nothing, 30, Nothing, 5]
+            t.at 'Price' . to_vector . should_equal [22.3, 32, 43.2, 54, 31, Nothing]
+
+        Test.specify "should read the specified sheet by name and properly handle dates"
+            t = read_method file sheet='Dates'
+            t.columns.map .name . should_equal ['Student Name', 'Enrolment Date']
+            t.at 'Enrolment Date' . map .day . to_vector . should_equal [2, 26, 4, 24, 31, 7]
```
Comment (radeusgd) [inline test/Table_Tests/src/Spreadsheet_Spec.enso]: While I like the conciseness, I'm a bit hesitant about the fact that both of these tests actually test two separate features, wouldn't it be cleaner to separate them? Thus if something breaks, we will know what was broken.

File: test/Table_Tests/src/Spreadsheet_Spec.enso [hunk 7]
```diff
@@ -0,0 +1,30 @@
+from Standard.Base import all
+import Standard.Test
+import Standard.Table
+
+spec_fmt header extension read_method =
+    file = Enso_Project.data / ('multisheet.' + extension)
+    Test.group header <|
```
Comment (radeusgd) [inline test/Table_Tests/src/Spreadsheet_Spec.enso:7]: I would suggest adding tests for loading empty sheets, as the logic for that in the Reader is not that trivial, so it would be good for the tests to check it.

**Review by radeusgd — CHANGES_REQUESTED** [review]
Looks good to me, but I'm worried about duplicated column names when reading with `has_header = true`. I would suggest adding some kind of deduplication or at least a warning + mention in the docs about it. I see that the CSV module most likely also has this issue, so if possible it would be good to update this.

If this is out of scope of this PR, maybe at least a separate issue for this could be created, but I think _at least_ a mention in the docs for now would be good. Otherwise it may be very confusing for someone trying to load a spreadsheet with duplicated header names.

File: distribution/lib/Standard/Table/0.1.0/src/Io/Spreadsheet.enso (hunk 3 above)
[the remark this answers is not in the document]
Comment (kustosz) [reply distribution/lib/Standard/Table/0.1.0/src/Io/Spreadsheet.enso:40]: I'm confused. This API here is exactly the same as for csv. You don't write `file.read.read_xlsx` you just write `file.read_xlsx`. I don't understand this point at all.

File: distribution/lib/Standard/Table/0.1.0/src/Io/Spreadsheet.enso [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+         example_xlsx_to_table = Examples.xlsx.read_xlsx
+
+   > Example
+     Read the sheet named `Dates` from an XLSX and convert it to a table.
+         import Standard.Table
+         import Standard.Examples
+
+         example_xlsx_to_table = Examples.xlsx.read_xlsx sheet='Dates'
+File.File.read_xlsx : Integer | Text | Nothing -> Boolean -> Text -> Table
```
[the remark this answers is not in the document]
Comment (kustosz) [reply distribution/lib/Standard/Table/0.1.0/src/Io/Spreadsheet.enso:38]: I did not, why would I? You read an XLS file, not read a file as an XLS.

File: distribution/lib/Standard/Table/0.1.0/src/Io/Spreadsheet.enso [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+
+         example_xlsx_to_table = Examples.xls.read_xls
+
+   > Example
+     Read the sheet named `Dates` from an XLS and convert it to a table.
+         import Standard.Table
+         import Standard.Examples
+
+         example_xls_to_table = Examples.xls.read_xls sheet='Dates'
+File.File.read_xls : Integer | Text | Nothing -> Boolean -> Text -> Table
```
[the remark this answers is not in the document]
Comment (kustosz) [reply distribution/lib/Standard/Table/0.1.0/src/Io/Spreadsheet.enso:109]: There is supposed to be a generic `read` method that allows reading any common format, we can have that there. This here should be more predictable.

File: test/Table_Tests/src/Spreadsheet_Spec.enso (hunk 6 above)
Comment (kustosz) [reply test/Table_Tests/src/Spreadsheet_Spec.enso]: We know that – it shows up twice in the report. Tests are still code, and copy-pasting a large block of code is usually bad.

File: distribution/lib/Standard/Table/0.1.0/src/Io/Spreadsheet.enso (hunk 9 above)
[the remark this answers is not in the document]
Comment (radeusgd) [reply distribution/lib/Standard/Table/0.1.0/src/Io/Spreadsheet.enso:109]: Sounds great

File: distribution/lib/Standard/Table/0.1.0/src/Io/Spreadsheet.enso (hunk 8 above)
[the remark this answers is not in the document]
Comment (radeusgd) [reply distribution/lib/Standard/Table/0.1.0/src/Io/Spreadsheet.enso:38]: You have a file `f` and `f.read_xlsx` reads `f` as XLS, so I would say that the converse is true, but I don't know, if this is the convention then I'm not strongly against it.

File: std-bits/table/src/main/java/org/enso/table/util/NameDeduplicator.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class NameDeduplicator {
+  public static List<String> deduplicate(List<String> names) {
+    return deduplicate(names, " ");
+  }
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/util/NameDeduplicator.java:11]: If we already have this mechanism, could we maybe use it in CSV loading too?

**Review by radeusgd — APPROVED** [review]
Looks good to me now. If you could re-use this deduplication in CSV reading that would be awesome!

One important note:
You have an entry containing weird characters in the file `/tools/legal-review/Table/org.apache.pdfbox.pdfbox-2.0.22/copyright-keep`, it makes the resulting NOTICES file be detected as a binary file.

The entry is

```
��targ��8rdesc ISO Coated v2 300% (basICColor)textbasICColor CMYKick v1.2 - Copyright (c) 2006-2007 Color Solutions, All Rights Reserved.
```

I would suggest moving this entry into `copyright-ignore` (you can just cut-and-paste the whole line, it should work) and manually adding an entry to `copyright-add` with contents: `Copyright (c) 2006-2007 Color Solutions, All Rights Reserved.`. This way we will preserve this copyright entry but get rid of these funky characters that confuse file readers.
